crates/thread_aware/src/cell.rs

```diff
@@ -456,21 +459,25 @@ impl<T, S: Strategy> ThreadAware for Trc<T, S> {
 
             let value = data;
 
-            let old_data = self
-                .storage
-                .write()
-                .expect("Failed to acquire write lock")
-                .replace(destination, Arc::<T>::clone(&value));
+            if let MemoryAffinity::Pinned(destination) = destination {
+                let old_data = self
+                    .storage
+                    .write()
+                    .expect("Failed to acquire write lock")
+                    .replace(destination, Arc::<T>::clone(&value));
 
-            assert!(old_data.is_none(), "Data already exists for the destination affinity");
+                assert!(old_data.is_none(), "Data already exists for the destination affinity");
```

Comment:
This issue had been there before, but this is a race condition - we'll panic in case two clones of this Trc try to be relocated to the same destination at the same time.